Initial (very rudimentary) flexbox implementation.

zentner-kyle opened the pull request:
This "flexbox implementation" is essentially a copy-and-paste / meld of code from various Flow types.

All of the flexbox related CSS properties should be parsed correctly. All of the children of the flex container (flex items) are (correctly) made into unfloated blocks. If the flexbox is a row or reverse-row, the total inline size is evenly distributed among the children, and they are laid out horizontally. Otherwise, the flexbox acts like a block with unfloated children.

I'd like to improve upon this code in a later pull request.

Closes #6636. I think?

r? @pcwalton

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7154)

<!-- Reviewable:end -->

Commit:
Create FlexFlow's for 'display: flex' nodes.

This commit doesn't implement any flexbox behavior at all.
It just constructs FlexFlow's, which act just like the BlockFlow from
which they "inherit."

Commit:
Parse flex-direction CSS property.

Commit:
Implement FlexFlow::assign_inline_sizes().

Commit:
Implement FlexFlow::assign_block_size().

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @metajack (or someone else) soon.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify layout code, but no reftests are modified. Please consider adding a reftest!

Commit:
Add reftests for basic flexbox functionality.

These tests (flex_row_direction.html and flex_column_direction.html) are
correct, but only test the small amount of current flexbox
functionality.

pcwalton commented:
Wow! I'll take a look at this tomorrow.

zentner-kyle commented:
Thanks for the feedback!

I think I've addressed the above comments? There weren't any on the later changes.

Yes, flex-done.rs was accidentally included.

Is rebasing mid-review like this an acceptable workflow?

pcwalton commented:
Yup, works for me. I'll finish the review ASAP.

pcwalton commented on components/layout/display_list_builder.rs:
```diff
@@ -1866,6 +1867,23 @@ impl ListItemFlowDisplayListBuilding for ListItemFlow {
+    fn build_display_list_for_flex(&mut self,
+                                   display_list: Box<DisplayList>,
+                                   layout_context: &LayoutContext) {
+        // Draw the rest of the block.
```
There may not be any need for this function—just call `build_display_list_for_block` directly in `FlexFlow::build_display_list`. Unless there will actually need to be special painting behavior for flexbox in the future?

pcwalton commented on components/layout/flex.rs:
```diff
@@ -0,0 +1,445 @@
+
+        let main_mode = match flex_style(&fragment).flex_direction {
+            flex_direction::T::row_reverse    => Mode::Inline,
+            flex_direction::T::row            => Mode::Inline,
```
nit: you can use or-patterns (`flex_direction::T::row_reverse | flex_direction::T::row => ...`) to avoid a little bit of code duplication here

pcwalton commented on components/layout/flex.rs:
```diff
@@ -0,0 +1,445 @@
+            main_mode: main_mode
+        };
+
+        this
```
No need for the `this` temporary; just write `BlockFlow { ... }` as the last expression.

pcwalton commented on components/layout/flex.rs:
```diff
@@ -0,0 +1,445 @@
+
+    // TODO(zentner): This function should actually flex elements!
+    // Currently, this is the core of TableRowFlow::assign_block_size() with
+    // float related logic stripped out.
```
Does `TableRowFlow` actually have any float related logic?

zentner-kyle replied:
Surprisingly, yes. However, it's relatively minimal (mostly a call to `kid.place_float_if_applicable()`).

pcwalton commented:
Layout looks good modulo a few nits. I'm going to bump the `properties.rs.mako` review over to @SimonSapin 

r? @SimonSapin

SimonSapin commented:
New properties and values should be marked "experimental". That should be easy for longhand properties (adding `experimental=True` at the right place), but for longhands (`flex: …`) or values (`display: flex`) you may need to add some Mako template support, or just call `::util::opts::experimental_enabled()` directly.

jdm commented:
I thought we only cared about marking things experimental that are not actually standardized and therefore shouldn't be exposed to regular web content, not merely incomplete things.

metajack commented:
This came up in the meeting a few weeks ago. Either @SimonSapin or @mbrubeck had some changes they planned to put behind `-e` until they were more fully baked.

SimonSapin commented:
Maybe we can rename "experimental" or have more than one flag, but the idea is that the barrier to entry for pref’ed-off things can be lower than otherwise. You can land piece-by-piece unfinished things that can panic (assertion errors, `unimplemented!()`, …) or give arbitrarily wrong results.

(I’m not saying at all that this PR is bad! I haven’t looked in details yet. But it doesn’t need to get everything right the first time.)

SimonSapin commented on components/style/properties.mako.rs:
```diff
@@ -258,6 +258,75 @@ pub mod longhands {
+                    if !(${condition}) {
+                        return Err(())
+                    }
+                    Ok(SpecifiedValue::Specified(value as ${type}))
```
Does this warn, when `type` is i32? (Trivial cast, or something like that.)

zentner-kyle replied:
That would make sense, but I don't think redundant / trivial casts ever do warn. At least, there aren't any warnings when I build this file using `./mach build`, and http://is.gd/9aTZQ1 doesn't show any warnings when compiling.

SimonSapin replied:
Ok.

SimonSapin commented on components/style/properties.mako.rs:
```diff
@@ -4783,6 +4852,92 @@ pub mod longhands {
+    // Flex item properties
+    ${integer_type_or_auto("order", "i32", "Specified(0)")}
+    ${integer_type_or_auto("flex-grow", "u32", "Specified(0)", "value >= 0")}
+    ${integer_type_or_auto("flex-shrink", "u32", "Specified(1)", "value >= 0")}
```
Does the layout code support all of these? It’s better not to parse properties or values for which we don’t have any support. (And adding `@supports` rules will make this more important.)

zentner-kyle replied:
Right now, almost none of these are supported (with the exception of `flex-direction`). However, I do intend to add support for most of them when I can. Should I remove them, or somehow feature gate them (besides using `experimental=True`)?

Also, I just realized that `align-self` is not in the right location in this list, it should be with the other flex item properties.

SimonSapin replied:
Yes, please remove the parsing of unsupported properties. Especially since they are one-line macro calls, they’re easy to add back when support is added.

SimonSapin commented on components/style/properties.mako.rs:
```diff
@@ -5444,6 +5599,40 @@ pub mod shorthands {
+        let flex_basis = if flex_basis.is_ok() {
+            flex_basis
+        } else {
+            flex_basis::parse(context, input)
```
Each of these four `foo::parse()` calls needs to be wrapped in `input.try(|input| …)` to avoid silently ignoring part of the input when they return `Err`.

SimonSapin commented:
(Done reviewing `properties.mako.rs` up to 47a6a0e82cff46324a26cb047c02ba8d41bc6bbe.)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7222) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
Very excited to see this initial PR.

For the record, here is a basic HTML template showing how we use flexbox model in browser.html: https://github.com/servo/servo/issues/7304 (works well in Firefox 43, appears to crash with this PR (panic: `attempted to divide by zero`)).

Hope this will help for future work on flexbox model.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7312) made this pull request unmergeable. Please resolve the merge conflicts.

zentner-kyle commented:
This is effectively merged due to #7312, so I'm closing this.

Thanks Simon and Patrick!
